packages/good-design/src/advanced/customswitch/CustomSwitch.tsx

```diff
@@ -0,0 +1,206 @@
+  )
+}
+
+//todo: add icon list (optional)
+export const CustomSwitch = ({list, switchListCb}:{list:string[], switchListCb: () => void}) => {
+  const [styleLeft, setStyleLeft] = useState<any>({
+    show: false,
+    style: hide
+  })
+  const [styleRight, setStyleRight] = useState<any>({
+    show: false,
+    style: hide
+  })
+
+  const [sourceList, setSourceList] = useState<string[]>(list);
+  const [targetList, setTargetList] = useState<string[]>(list.slice().reverse());
+
+  const toggleList = (left?: boolean) => {
+    const side = left ? [styleLeft, setStyleLeft ] : [styleRight, setStyleRight ]  
+    !side[0].show ? side[1]({ show: true, style: show }) : side[1]({ show: false, style: hide })
+  }
+
+  const switchSelect = () => {
+    setSourceList(targetList);
+    setTargetList(sourceList);
+    switchListCb();
+  }
+
+  const selectFromList = (index: number, isLeft: boolean) => {
+    const sides = [sourceList, targetList]
+    const selectedSide = isLeft ? sides.splice(0, 1)[0] : sides.splice(1,1)[0]
+    const altSide = sides[0]
+    const selected = selectedSide.splice(index, index)[0]
+    selectedSide.unshift(selected);
+    if (altSide.indexOf(selected) === 0) {
+      const altSelected = altSide.splice(index, index)[0]
+      altSide.unshift(altSelected)
+    }
+    toggleList(isLeft);
+    switchListCb(); //Todo: refactor to handle list with more then 2 values
+  }
+
+  return (
+    <View height="16" display="flex" flexDirection="row" justifyContent="flex-start" alignItems="flex-start">
+      <Box display="flex" alignItems="center" justifyContent="center" flexDirection="column">
+        {
+          sourceList.map((chain, index) => (
+            <SelectListItem
+              key={index}
+              chain={index === 0 ? sourceList[0] : chain}
+              styles={styleLeft.style}
+              press={() => index === 0 ? toggleList(true) : selectFromList(index, true)}
+              isListItem={index !== 0}
+            />
+          ))
+        }
+      </Box>
+      <Box w="100px" height="64px" pl="3" pr="3" display="flex" justifyContent={"center"} alignItems="center">
+        <Pressable onPress={switchSelect}>
+          <Icon
+            minWidth="8"
+            size="xl"
+            w="6"
+            h="6"
+            display="flex"
+            alignSelf="center"
+            justifyItems="center"
+            pl="1.5"
+            fill="none"
+            viewBox="0 0 24 24">
+              <SwitchIcon />
+            </Icon>
+        </Pressable>
+      </Box>
+      <Box display="flex" alignItems="center" justifyContent={"center"} flexDirection="column">
+        {
+          targetList.map((chain, index) => (
+            <SelectListItem
+              key={'target'+index}
+              chain={index === 0 ? targetList[0] : chain}
+              styles={styleRight.style}
+              press={() => index === 0 ? toggleList(false) : selectFromList(index, false)}
```

Comment:
1. no inline callbacks !
2. no index as list keys


```javascript
// CustomSwitch component
const onUncheckList = useCallback(() => toggleList(false), [toggleList])
const onUncheckItem = useCallback(chain => selectFromList(chain.chainId, false), [toggleList]) 
// use unique id instead of idx. chainId is my guess, you could replace it
// with any other unique field from chain object

// ...
// use unique id instead of idx
<SelectListItem 
  key={`target${chain.chainId}`} 
  chain={index ? chain : first(targetList)} 
  onPress={index ? onUncheckItem : onUncheckList}
  isListItem={!!index}
  // ...

// SelectListItem component
const onItemPress = useCallback(() => onPress(chain), [chain, onPress]); // pass this handler down
```

Reply:
Seems to work... but in this flow there is a hook rule we break where the order is different (after selecting from list)

Is that because the two different callbacks are run (first open list > second select item)?

Reply:
we don't break any hook rules
both callback are declared each render
hooks count & order unchanged